fix: EC2 agent fails open when cosign or the verification key is unavailable, and no test covers it

## Problem

The EC2 agent **fails open for authenticity**: if cosign is not on `PATH`, or the public key cannot be fetched, the agent constructs a nil verifier and continues booting with SHA-256 checking alone. The single most security-relevant branch in the agent is also the one branch the test suite does not exercise.

### The fail-open

`cmd/strata-agent/main.go:77-90` — `newCosignVerifier()` returns nil on either of two conditions:

- `exec.LookPath("cosign")` returns an error → `return nil`
- `fetchPublicKey(ctx) == ""` → `return nil`

Both branches log "skipping bundle verification". The intent is stated in the doc comment at `cmd/strata-agent/main.go:74-76`: "degrades gracefully (SHA256 integrity is still enforced)".

`internal/agent/agent.go:269-272`, inside `verifyBundles`:

```go
if a.cfg.Verifier == nil || a.cfg.BundleFetcher == nil {
	return nil
}
```

So step 3 of `Run` succeeds vacuously and steps 4–6 (mount, configure, signal) proceed. `[READ]`

The primitive is fail-closed; only the caller is fail-open. `internal/trust/verify.go:24-68` requires *all* of SHA256 match, non-empty bundle, non-URI bundle path, parseable bundle, `HasRekorEntry()`, and `v.Verify()` before returning nil. That code is correct. It is simply not reached.

### The missing test

```
$ grep -rn "Verifier: *nil" --include='*_test.go' .
(no output)

$ go test ./internal/agent -v
TestRun_HappyPath
TestRun_NoLayers
TestRun_AcquireFails
TestRun_FetchFails
TestRun_SHA256Mismatch
TestRun_MountFails
```

Six tests, no verifier-absent case. A change that turned this fail-open into a fail-closed — **or the reverse** — would pass CI silently. `internal/agent` sits at 58.2% statement coverage; this branch is in the other 41.8%.

## Measurement limit, stated explicitly

This is graded `[READ]`. The nil guard is two lines and reading it settles what the code *does*. What reading cannot show is a real EC2 instance booting fail-open. That would need an instance with the `strata-builder` instance profile, a reachable `strata-registry` bucket, and no cosign on `PATH`.

The cheap closure is not an EC2 instance. It is the unit test that does not exist — which is why the gap exists at all.

## Contrast: the agent gets the *hash* path right

Worth stating so the fix does not regress it. `internal/agent/agent.go:224-242` (`fetchAndVerifyLayers`) calls `Fetch` and then **unconditionally** hashes the returned path and compares to `layer.SHA256`; the cache-hit early return lives *inside* `cmd/strata-agent/s3_fetcher.go:71-80`, below the hash. That is the correct structure, and it is why the agent does not share the `strata run` cache bypass filed separately in this cohort.

## What closes this

`agent.Config{Verifier: nil}` makes `Run` return an error, and a test asserts it. Concretely:

- `TestRun_NilVerifier` constructing an `agent.Config` with `Verifier: nil` and asserting `Run` fails
- the same for `BundleFetcher: nil`
- `newCosignVerifier` returning an error rather than a nil interface value, so the failure surfaces at construction

If fail-open is a deliberate, accepted operational trade-off for boot availability, then it needs to be an explicit opt-in (a flag or instance tag, defaulting to closed) and `STRATA.md:288-290`'s "unconditional" claim has to go. A silent default fail-open with no test is not a trade-off, it is an accident waiting to be committed.

---

*Filed from the 2026-08-20 recon session, grading an external deep review against clone `64d28bd` (go1.26.7 darwin/arm64). Covers review assertions **A1, A2** (both confirmed) and finding **F14** (the missing test — not in the review). `[RUN]` = established by executing something; `[READ]` = established by reading the cited line.*

## The pair decision: **two PRs**, #55 first, neither depending on the other landing

The v0.23.0 working order groups #55, #56 and #63 into one slot. Decided on independent review and independent revert, not on subject similarity.

**Independent revert — yes, structurally.** Two separate `main` packages with no import path between them:

```
$ git log -1 --format=%h -- cmd/strata/run.go          # 415b90e   -> #55's file
$ git log -1 --format=%h -- cmd/strata-agent/main.go   # 13469df   -> #56's file
$ grep -rn "cmd/strata-agent" cmd/strata/ ; grep -rn "cmd/strata\"" cmd/strata-agent/    # no output
```

Zero file overlap. Reverting either restores exactly one binary's behaviour.

**What each side changed that the other consumes — checked separately from conflict, because a conflict reports collision, not interaction.** Both callers consume `internal/trust` (`VerifyLayer`, `Verifier`, `ParseBundle`). The interaction hazard is real *if* either fix modifies `internal/trust`. Neither needs to:

- #55 reuses `loadLocalBundle` from `cmd/strata/verify.go` (package `main`, added under #59) and calls `trust.VerifyLayer` unchanged.
- #56's reachable half changes `newCosignVerifier`'s signature in `cmd/strata-agent/main.go` only.

If either PR ends up touching `internal/trust`, that invalidates this half of the decision and the two must be re-examined as one. I will state at PR time whether it held.

**Independent review — yes, and this is the load-bearing reason.** The two ask a reviewer different questions with different risk calculus:

- #55: should a user-space command refuse to mount unverified content? Failure mode of getting it wrong: a researcher cannot start work, sees an error, passes a flag.
- #56: should an EC2 instance refuse to **boot** when cosign is missing? Failure mode: instances fail at boot, in a fleet, where the operator sees a dead instance and pays for it. Boot-availability trade-offs are a different argument from command ergonomics.

Bundled, a reviewer gives one verdict over two risk profiles, and the cheaper one (#55) carries the expensive one (#56) through on momentum. That is the thing to avoid, and it is what makes this two PRs rather than one.

**What a reviewer would lose in a single PR** — the ability to accept enforcement in `strata run` while rejecting boot-refusal on EC2. Those are separable positions and a plausible reviewer holds exactly that combination.

**Order and dependency.** #55 first. There is no code dependency in either direction — #56 does not need #55 landed, and vice versa. #55 goes first because it is completable in full under this session's constraints, whereas #56 is only partly completable (below), so #56's PR carries a decision request and I would rather that arrive attached to a PR that already contains everything it can.

---

## Gate, answered before any code: does any CI job execute this code path?

This one splits, and the split is the finding.

### The agent binary's constructor: not reached at all

```
$ grep -n "cmd/strata-agent\|strata-agent" .github/workflows/ci.yml
(no output)

$ grep -rn "newCosignVerifier\|fetchPublicKey" cmd/strata-agent/*_test.go
(no output)

$ go test -covermode=atomic -coverprofile=cov.out ./cmd/strata-agent/... && go tool cover -func=cov.out | grep -E "newCosignVerifier|fetchPublicKey"
cmd/strata-agent/main.go:77:  newCosignVerifier  0.0%
cmd/strata-agent/main.go:92:  fetchPublicKey     0.0%
```

`cmd/strata-agent` appears nowhere in the workflow. No job builds or runs the agent binary; `go test ./...` compiles it and stops there.

### The library's nil guard: reached constantly, and asserted by nothing

This is worse than uncovered, and it is worth being precise about why. From the same profile at `internal/agent/agent.go:270`:

```
internal/agent/agent.go:269.104,270.57   1  3     <- the guard is evaluated 3x
internal/agent/agent.go:270.57,272.3     1  3     <- the fail-open `return nil` is TAKEN 3x
internal/agent/agent.go:275.2,276.26     2  0     <- everything after it: never runs
```

`verifyBundles` reads 4.3%. `internal/agent/agent_test.go` never mentions `Verifier` or `BundleFetcher` at all, so all six `TestRun_*` cases construct a `Config` leaving both nil, take the fail-open branch, and pass.

So "does CI execute the path?" answers **yes** for the most security-relevant branch in the agent — and the answer is worthless. Coverage tooling reports the line green. A change inverting this guard, or re-breaking it, is invisible to CI. That is the same shape as the sentinel problem recorded on #73: *execution is not assertion*, exactly as a sentinel firing is not proof the input could have failed differently.

Provenance: head `9a735ca7b11d4a1db8433bbbcd6a52c484c2d364`, tree `ee5f730428c844a52f66dda760696c17e57e3665`, dirty 0, `go1.26.7 darwin/arm64`. Source line under test: `internal/agent/agent.go:270` — `if a.cfg.Verifier == nil || a.cfg.BundleFetcher == nil {`.

---

## The part of this issue I cannot close this session, measured rather than guessed

This issue's stated closing condition is `agent.Config{Verifier: nil}` making `Run` return an error. **That cannot be done without modifying an inherited test file**, which this session's contract forbids and a `PreToolUse` hook enforces.

Measured, not predicted — in a detached worktree at `9a735ca` I inverted the guard to return an error and ran the suite:

```
$ git worktree add --detach /tmp/s08/probe-invert 9a735ca
# guard at internal/agent/agent.go:270 changed to return an error
$ go test ./internal/agent/...
--- FAIL: TestRun_HappyPath (0.00s)
    agent_test.go:82: Run: agent: verifying bundles: agent: refusing to boot: verification not configured
--- FAIL: TestRun_NoLayers (0.00s)
    agent_test.go:119: Run with no layers: ...
--- FAIL: TestRun_MountFails (0.00s)
    agent_test.go:253: Run error should wrap ErrNotSupported: got ...
FAIL
```

Three inherited tests in `internal/agent/agent_test.go` assert that a nil verifier boots successfully. They are not wrong about today's code; they encode the behaviour this issue calls a defect. Making `Run` fail closed requires editing them.

Per the hazard rule — *if a correct fix appears to require loosening a check or special-casing the fixture, the fixture is what is wrong* — I am arguing it before any change rather than after: **the inherited tests are what is wrong here**, because `TestRun_HappyPath` calls a boot with authenticity checking disabled a happy path. But "the fixture is wrong" does not grant me permission to edit it under a contract that forbids exactly that, so this is a decision, not a fix.

`TestRun_NoLayers` failing is a separate, smaller signal worth recording: the guard fires before the `len(toVerify) == 0` check, so a boot with nothing to verify is refused too. A correct fail-closed version should test "is there anything to verify" first.

**So this PR closes the reachable half and leaves the library half open, explicitly.** After the fix, `newCosignVerifier` returns an error and `main` treats it as fatal, so no shipped caller can hand `internal/agent` a nil `Verifier` — `BundleFetcher` is always non-nil (`cmd/strata-agent/main.go:51`), so `Verifier` is the only way in. The library-level permissiveness survives as unreachable-from-production, which is the same shape as #59's outcome at resolver stage 7, and it will be filed with that framing rather than left implied.

## What I am going to build

`newCosignVerifier(ctx) (trust.Verifier, error)` — an error, never a nil interface value. `main` exits rather than booting unverified. The opt-out is explicit and defaults to closed, per this issue's own condition: an environment variable the operator must set, which logs loudly that authenticity checking is off.

## Forecast

**1. Files.**
- `cmd/strata-agent/main.go` — `newCosignVerifier` returns `(trust.Verifier, error)`; call site at line 52 handles it; explicit opt-out.
- `cmd/strata-agent/cosign_verifier_test.go` — **new file**.
- `CHANGELOG.md` — behaviour change, per consumer.
- **Not** `internal/agent/agent.go`, and **not** `internal/agent/agent_test.go`.

**2. What will be green.** `go test -race ./...` clean; `newCosignVerifier` above 0.0%, from a test that manipulates `PATH` so cosign is genuinely absent.

**3. The observable that closes the reachable half.** With cosign off `PATH`, `newCosignVerifier` returns an error naming the missing prerequisite; with the opt-out set it returns a nil verifier *and* says so. The must-reject material is the absent-cosign case, which today returns a nil verifier and a log line.

**4. Two predicted failures about the system.**
- (a) `fetchPublicKey` calls `awsconfig.LoadDefaultConfig` and S3, so a test that gets as far as the key fetch will either hit the network or hang on credential resolution in CI. I predict the first cut of the test reaches it and I have to structure the check so the cosign-absent case short-circuits before any AWS call.
- (b) `exec.LookPath("cosign")` consults `PATH` at call time, so a test that sets `PATH=""` will interact with `t.Setenv` and parallelism. I predict a `t.Parallel()`-related flake or a linter complaint about `t.Setenv` in a parallel test.

**5. What makes me abandon and come back for a decision.** Already triggered, above: the library half. I am not editing an inherited test file to get it, and I am not weakening the fix to fit the tests.

**6. Predicted failure about me** — scored separately, not already recorded on an issue.

> Having found the "executed 3 times, asserted never" result, I will over-reach on it: I will treat the coverage profile as proof that *only* these three tests take the branch, and state a count of affected tests from the profile's hit count rather than from the failing test list. The hit count is 3 and the failing tests are 3, which makes the wrong inference look confirmed. They measure different things — one counts branch executions, the other counts tests whose assertions depend on it.

Falsifiable: the numbers happen to agree here, so if I cite the profile's `3` as evidence about *tests*, the prediction hit even though the number is right.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,53 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
 ### Security
+- **`strata-agent` refuses to boot when it cannot verify layer authenticity**
+  (#56). `newCosignVerifier` returned a nil `trust.Verifier` when cosign was not
+  on `PATH` or the public key could not be fetched from S3, and
+  `internal/agent.verifyBundles` treats a nil verifier as nothing to do — so an
+  EC2 instance silently downgraded from authenticity checking to hash checking.
+  "cosign is not installed" and "the operator accepted the risk" were the same
+  value, and the second was never asked for.
+
+  `newCosignVerifier` now returns `(trust.Verifier, error)` and names the missing
+  prerequisite. The boot stops, signalling failure to EC2 first so the instance
+  is diagnosable rather than looking hung. The opt-out is explicit and defaults
+  to closed: `STRATA_AGENT_ALLOW_UNVERIFIED=1` boots without authenticity
+  verification and logs that it has done so. Unrecognised values — including
+  typos like `ture` — leave verification on.
+
+  A nil verifier is now returned **only** with the opt-out set. That pairing, not
+  "never nil", is the invariant, and it is what the new tests assert.
+
+  What is given up by opting out is *authenticity*, not integrity: SHA-256
+  against the lockfile is still enforced on every path. The distinction matters
+  because a lockfile is exactly the artifact an attacker would supply, so it
+  cannot vouch for who produced the layers it pins.
+
+  **What changes per consumer**, enumerated from
+  `grep -rn 'agent.Config{' --include='*.go' .` and
+  `grep -rn 'internal/agent' --include='*.go' . | grep -v '^./internal/agent'`:
+  - **`strata-agent` on an instance without cosign, or without registry access
+    to `build/keys/cosign.pub`, now fails to boot** where it previously booted
+    with authenticity checking off. This is the intended breaking change.
+    Operators who relied on the old degradation must set
+    `STRATA_AGENT_ALLOW_UNVERIFIED=1`. The AMI build is the place to check: if
+    cosign is not baked in, every instance was taking the old path.
+  - `internal/agent` is **unchanged** — deliberately. `verifyBundles`'s
+    `Verifier == nil` guard is still fail-open, and inverting it fails three
+    inherited tests in `internal/agent/agent_test.go` that construct a boot with
+    verification disabled and call it a happy path. That is a reviewed decision,
+    not a fix to slip in, and it is filed as such. What *is* closed is the
+    deployed hole: `internal/agent` is an internal package with exactly one
+    non-test `agent.Config` construction (`cmd/strata-agent/main.go`), where
+    `BundleFetcher` is never nil, so after this change no shipped caller can
+    reach the library fail-open.
+  - `STRATA.md`'s layer-verification section claimed verification "is
+    unconditional — there is no flag to skip verification". That was false in the
+    worse direction: verification was skippable without a flag. It now documents
+    both explicit opt-outs.
+  - `pkg/strata` and every other command are **unchanged**: nothing outside
+    `cmd/strata-agent` imports `internal/agent`.
 - **Rekor entry verification compares the entry to the bundle** (#59).
   `RekorHTTPClient.VerifyEntry` took a bundle, discarded it (`_ = bundle`), and
   returned success for any log index the server could find an entry at. That is a

--- a/STRATA.md
+++ b/STRATA.md
@@ -286,7 +286,27 @@ entry commits the layer's identity to a public, append-only log.
 
 The agent verifies the cosign bundle against the Rekor transparency log before mounting
 any layer. SHA256 of the pulled squashfs is verified against the manifest. Unsigned layers
-will not mount. This is unconditional — there is no flag to skip verification.
+will not mount.
+
+Verification is required, not unconditional, and the difference is worth stating because
+this paragraph previously claimed the stronger thing. Two escape hatches exist, and both
+are explicit:
+
+- `strata run` requires `--key` and refuses to mount without it; `--no-verify` mounts
+  unverified layers and says so on stderr (#55).
+- `strata-agent` refuses to boot when cosign or the public key is unavailable. Setting
+  `STRATA_AGENT_ALLOW_UNVERIFIED=1` in the instance's user-data or unit file lets it boot
+  without authenticity verification, logging that it has done so (#56).
+
+What the earlier "there is no flag to skip verification" sentence obscured is that
+verification was skippable *without* a flag: the agent returned a nil verifier whenever
+cosign was missing, and the library treated a nil verifier as nothing to do. A documented
+opt-out that defaults to closed is a weaker claim than the one this paragraph used to make
+and a stronger property than the code used to have.
+
+In every case SHA256 integrity against the lockfile is still enforced. What an opt-out
+gives up is *authenticity* — evidence about who produced the layers — which is precisely
+what a supplied lockfile cannot vouch for on its own.
 
 ### Lockfile signing
 

--- a/cmd/strata-agent/cosign_verifier_test.go
+++ b/cmd/strata-agent/cosign_verifier_test.go
@@ -1,0 +1,282 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/scttfrdmn/strata/internal/trust"
+)
+
+// Tests for the agent's verifier construction (#56).
+//
+// The defect was that "cosign is not installed" and "the operator accepted the
+// risk" were the same value — a nil trust.Verifier — so an EC2 instance
+// downgraded from authenticity checking to hash checking without anyone asking
+// it to. Before this change `newCosignVerifier` had 0.0% coverage: it was called
+// only from main(), so nothing in CI executed the most security-relevant branch
+// in the binary.
+//
+// The property under test is therefore not "a verifier is built". It is that a
+// nil verifier is returned if and only if the operator explicitly opted out.
+
+func cosignPresent(string) (string, error) { return "/usr/local/bin/cosign", nil }
+
+// cosignAbsent reproduces exec.LookPath's failure shape.
+func cosignAbsent(string) (string, error) {
+	return "", errors.New(`exec: "cosign": executable file not found in $PATH`)
+}
+
+func keyFetched(context.Context) string     { return "/tmp/strata-cosign-test.pub" }
+func keyUnavailable(context.Context) string { return "" }
+
+// TestNewCosignVerifier_NamesTheMissingPrerequisite covers the two refusal
+// directions and the one success direction.
+//
+// Both refusals must be reachable without cosign installed and without AWS
+// credentials, which is the state of CI — that reachability is most of what this
+// issue was about.
+func TestNewCosignVerifier_NamesTheMissingPrerequisite(t *testing.T) {
+	tests := []struct {
+		name       string
+		prereqs    verifierPrereqs
+		wantReason string
+	}{
+		{
+			name:       "cosign missing",
+			prereqs:    verifierPrereqs{lookPath: cosignAbsent, fetchKey: keyFetched},
+			wantReason: "cosign not found on PATH",
+		},
+		{
+			// Reported as the key, not as cosign: both prerequisites are
+			// consulted in order and this row proves the second one is
+			// reached rather than masked by the first.
+			name:       "cosign present, key unavailable",
+			prereqs:    verifierPrereqs{lookPath: cosignPresent, fetchKey: keyUnavailable},
+			wantReason: "could not fetch the cosign public key",
+		},
+		{
+			name:    "both prerequisites present",
+			prereqs: verifierPrereqs{lookPath: cosignPresent, fetchKey: keyFetched},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			v, err := newCosignVerifier(context.Background(), tt.prereqs)
+
+			if tt.wantReason == "" {
+				if err != nil {
+					t.Fatalf("want a verifier, got error: %v", err)
+				}
+				if v == nil {
+					t.Fatal("returned (nil, nil): a nil Verifier is a panic in trust.VerifyLayer, not a skip")
+				}
+				cv, ok := v.(*trust.CosignVerifier)
+				if !ok {
+					t.Fatalf("want *trust.CosignVerifier, got %T", v)
+				}
+				if cv.KeyRef != keyFetched(context.Background()) {
+					t.Errorf("verifier does not use the fetched key: KeyRef = %q", cv.KeyRef)
+				}
+				return
+			}
+
+			if err == nil {
+				t.Fatalf("want a refusal, got verifier %+v", v)
+			}
+			if v != nil {
+				t.Errorf("returned a verifier alongside an error: %+v", v)
+			}
+			if !strings.Contains(err.Error(), tt.wantReason) {
+				t.Errorf("refusal names the wrong prerequisite:\n got: %v\nwant substring: %s",
+					err, tt.wantReason)
+			}
+		})
+	}
+}
+
+// TestNewCosignVerifier_ReportsCosignBeforeTheKey: when both prerequisites are
+// missing, the operator must be sent to the binary and not to the key, because
+// without the binary the key is useless.
+//
+// This is the row that the two-refusal table above cannot express — it needs
+// both defects present at once, and it asserts which one wins.
+func TestNewCosignVerifier_ReportsCosignBeforeTheKey(t *testing.T) {
+	_, err := newCosignVerifier(context.Background(),
+		verifierPrereqs{lookPath: cosignAbsent, fetchKey: keyUnavailable})
+	if err == nil {
+		t.Fatal("want a refusal")
+	}
+	if !strings.Contains(err.Error(), "cosign not found on PATH") {
+		t.Errorf("want the cosign prerequisite reported first, got: %v", err)
+	}
+	if strings.Contains(err.Error(), "public key") {
+		t.Errorf("key fetch was attempted or reported despite cosign being absent: %v", err)
+	}
+}
+
+// TestResolveVerifier_NilVerifierOnlyWithAnExplicitOptOut is the invariant.
+//
+// A nil verifier is not an error value here — it is the opt-out result — so the
+// assertion cannot be "never nil". It is that the pairing (nil verifier, nil
+// error) occurs only when allowUnverified is set, and that every other
+// prerequisite failure is an error the boot stops on.
+func TestResolveVerifier_NilVerifierOnlyWithAnExplicitOptOut(t *testing.T) {
+	failing := []struct {
+		name    string
+		prereqs verifierPrereqs
+	}{
+		{"cosign missing", verifierPrereqs{lookPath: cosignAbsent, fetchKey: keyFetched}},
+		{"key unavailable", verifierPrereqs{lookPath: cosignPresent, fetchKey: keyUnavailable}},
+	}
+
+	for _, f := range failing {
+		t.Run(f.name+", opt-out not set: refuses", func(t *testing.T) {
+			v, err := resolveVerifier(context.Background(), f.prereqs)
+			if err == nil {
+				t.Fatalf("boot was allowed to continue without verification (verifier %+v)", v)
+			}
+			if v != nil {
+				t.Errorf("returned a verifier alongside an error: %+v", v)
+			}
+			if !strings.Contains(err.Error(), allowUnverifiedEnv) {
+				t.Errorf("refusal does not name the opt-out an operator would need: %v", err)
+			}
+		})
+
+		t.Run(f.name+", opt-out set: proceeds without a verifier", func(t *testing.T) {
+			p := f.prereqs
+			p.allowUnverified = true
+			v, err := resolveVerifier(context.Background(), p)
+			if err != nil {
+				t.Fatalf("explicit opt-out must not fail the boot: %v", err)
+			}
+			if v != nil {
+				t.Errorf("opt-out returned a verifier: %+v", v)
+			}
+		})
+	}
+
+	t.Run("prerequisites present: a verifier, regardless of the opt-out", func(t *testing.T) {
+		// The opt-out is permission to boot unverified, not an instruction to
+		// stop verifying. An operator who sets it on a fleet where cosign is
+		// present should still get verification.
+		for _, allow := range []bool{false, true} {
+			v, err := resolveVerifier(context.Background(), verifierPrereqs{
+				lookPath: cosignPresent, fetchKey: keyFetched, allowUnverified: allow,
+			})
+			if err != nil {
+				t.Fatalf("allowUnverified=%v: unexpected error: %v", allow, err)
+			}
+			if v == nil {
+				t.Errorf("allowUnverified=%v: verification was skipped although both prerequisites were present", allow)
+			}
+		}
+	})
+}
+
+// TestAllowUnverified_DefaultsToClosed drives the opt-out parser through the
+// values an operator can actually produce, including the ones that look
+// affirmative and are not.
+//
+// The unset case is first because it is the one production depends on and the
+// one a table of hand-set booleans would never reach.
+func TestAllowUnverified_DefaultsToClosed(t *testing.T) {
+	tests := []struct {
+		value string
+		want  bool
+	}{
+		{"", false}, // unset, and the production default
+		{"1", true},
+		{"true", true},
+		{"TRUE", true},
+		{"  yes  ", true}, // whitespace is trimmed
+
+		{"on", true},
+		{"0", false},
+		{"false", false},
+		{"no", false},
+		// Anything unrecognised leaves verification on. A typo in the value is
+		// not consent, and this is the direction of error a security default
+		// should have.
+		{"maybe", false},
+		{"ture", false},
+		{"2", false},
+		{"disabled", false},
+	}
+	for _, tt := range tests {
+		got := allowUnverified(func(string) string { return tt.value })
+		if got != tt.want {
+			t.Errorf("allowUnverified(%q) = %v, want %v", tt.value, got, tt.want)
+		}
+	}
+}
+
+// TestAllowUnverified_ReadsTheRightVariable: a parser that ignored its key would
+// pass every row above, because they all return the same value for any key.
+func TestAllowUnverified_ReadsTheRightVariable(t *testing.T) {
+	getenv := func(k string) string {
+		if k == allowUnverifiedEnv {
+			return "1"
+		}
+		return ""
+	}
+	if !allowUnverified(getenv) {
+		t.Errorf("allowUnverified does not read %s", allowUnverifiedEnv)
+	}
+
+	other := func(k string) string {
+		if k == "STRATA_AGENT_ALLOW_UNVERIFIED_LAYERS" {
+			return "1"
+		}
+		return ""
+	}
+	if allowUnverified(other) {
+		t.Error("a similarly-named variable enabled the opt-out")
+	}
+}
+
+// TestProductionPrereqs_ClosedByDefault reaches the default through the wiring
+// production uses, rather than by setting the field.
+//
+// This test exists because of a predicted failure recorded on #56 before the
+// code was written: that the closed default would be correct by construction and
+// asserted by nothing, since every other test here sets allowUnverified
+// explicitly. Reading the literal `false` back with one's eyes is not a check.
+func TestProductionPrereqs_ClosedByDefault(t *testing.T) {
+	// A getenv that reports every variable as unset — the state of a fresh
+	// instance whose user-data says nothing about verification.
+	unset := func(string) string { return "" }
+
+	p := productionPrereqs(unset)
+	if p.allowUnverified {
+		t.Error("the production default opts out of verification")
+	}
+	if p.lookPath == nil || p.fetchKey == nil {
+		t.Fatal("production prereqs are incompletely wired")
+	}
+
+	// And the consequence, through resolveVerifier rather than by inspecting
+	// the field: with cosign unavailable and nothing set, the boot stops.
+	p.lookPath = cosignAbsent
+	if _, err := resolveVerifier(context.Background(), p); err == nil {
+		t.Error("with nothing set and cosign absent, the boot was allowed to continue")
+	}
+}
+
+// TestProductionPrereqs_UsesTheRealEnvironment checks the wiring against
+// os.Getenv, which is what main passes. os.Setenv is safe here because the
+// variable is read once, synchronously, by the call under test.
+func TestProductionPrereqs_UsesTheRealEnvironment(t *testing.T) {
+	if productionPrereqs(os.Getenv).allowUnverified {
+		t.Fatalf("%s is set in this test environment; the default cannot be observed", allowUnverifiedEnv)
+	}
+
+	t.Setenv(allowUnverifiedEnv, "1")
+	if !productionPrereqs(os.Getenv).allowUnverified {
+		t.Errorf("productionPrereqs does not read %s from the environment", allowUnverifiedEnv)
+	}
+}

--- a/cmd/strata-agent/cosign_verifier_test.go
+++ b/cmd/strata-agent/cosign_verifier_test.go
@@ -267,6 +267,29 @@ func TestProductionPrereqs_ClosedByDefault(t *testing.T) {
 	}
 }
 
+// TestProductionPrereqs_RefusesWithTheRealLookPath closes the gap the injected
+// lookPath leaves: that the production wiring actually consults the real one.
+//
+// Every other test here substitutes cosignAbsent, which proves the decision
+// logic and nothing about what production calls. Stripping PATH makes the real
+// exec.LookPath fail — deterministically, whether or not cosign is installed on
+// the machine running this — and no S3 call happens, because the key fetch is
+// never reached.
+func TestProductionPrereqs_RefusesWithTheRealLookPath(t *testing.T) {
+	t.Setenv("PATH", "/nonexistent-strata-agent-test")
+
+	v, err := resolveVerifier(context.Background(), productionPrereqs(os.Getenv))
+	if err == nil {
+		t.Fatalf("cosign is unreachable and nothing was opted out, but the boot was allowed to continue (verifier %+v)", v)
+	}
+	if v != nil {
+		t.Errorf("returned a verifier alongside an error: %+v", v)
+	}
+	if !strings.Contains(err.Error(), "cosign not found on PATH") {
+		t.Errorf("refusal does not name the real prerequisite: %v", err)
+	}
+}
+
 // TestProductionPrereqs_UsesTheRealEnvironment checks the wiring against
 // os.Getenv, which is what main passes. os.Setenv is safe here because the
 // variable is read once, synchronously, by the call under test.

--- a/cmd/strata-agent/main.go
+++ b/cmd/strata-agent/main.go
@@ -12,11 +12,13 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
+	"fmt"
 	"io"
 	"log"
 	"os"
 	"os/exec"
 	"path/filepath"
+	"strings"
 	"time"
 
 	"github.com/aws/aws-sdk-go-v2/aws"
@@ -45,11 +47,26 @@ func main() {
 	fetcher := newS3LayerFetcher()
 	signaler := newEC2ReadySignaler()
 
+	// Resolve the verifier before anything else is built. A boot that cannot
+	// check authenticity is a boot that stops, and it stops here rather than
+	// six steps later with the layers already mounted.
+	verifier, err := resolveVerifier(ctx, productionPrereqs(os.Getenv))
+	if err != nil {
+		// Signal before dying. An instance that refuses to boot and says
+		// nothing is indistinguishable from an instance that hung, and the
+		// operator is paying for it either way. Signalling is best-effort;
+		// its failure is logged and must not make the refusal non-fatal.
+		if sigErr := signaler.SignalFailed(ctx, err); sigErr != nil {
+			log.Printf("strata-agent: could not signal boot failure: %v", sigErr)
+		}
+		log.Fatalf("strata-agent: %v", err)
+	}
+
 	a, err := agent.New(agent.Config{
 		Source:           newMetadataLockfileSource(),
 		Fetcher:          fetcher,
 		BundleFetcher:    fetcher,
-		Verifier:         newCosignVerifier(ctx),
+		Verifier:         verifier,
 		Signaler:         signaler,
 		PackageInstaller: agent.ExecPackageInstaller{},
 	})
@@ -71,24 +88,122 @@ func main() {
 	}
 }
 
-// newCosignVerifier checks that cosign is available on PATH and downloads the
-// public key from S3. Returns nil when either prerequisite is missing so the
-// agent degrades gracefully (SHA256 integrity is still enforced).
-func newCosignVerifier(ctx context.Context) trust.Verifier {
-	if _, err := exec.LookPath("cosign"); err != nil {
-		log.Printf("strata-agent: cosign not found on PATH; skipping bundle verification")
-		return nil
-	}
-	keyPath := fetchPublicKey(ctx)
-	if keyPath == "" {
-		log.Printf("strata-agent: could not fetch cosign public key; skipping bundle verification")
-		return nil
-	}
-	return &trust.CosignVerifier{KeyRef: keyPath}
+// allowUnverifiedEnv opts the agent out of authenticity verification. Unset —
+// or set to anything other than an affirmative value — means verification is
+// required and the agent refuses to boot without it.
+//
+// An environment variable rather than a flag because the agent is a systemd
+// unit with no meaningful argv, and because it matches the existing
+// STRATA_REGISTRY_BUCKET lever. Setting it is a recorded act in the instance's
+// user-data or unit file, which is the property that matters: the previous
+// behaviour was a downgrade nobody had to ask for.
+const allowUnverifiedEnv = "STRATA_AGENT_ALLOW_UNVERIFIED"
+
+// verifierPrereqs are the inputs the verifier decision consults.
+//
+// They are injected rather than called directly so that every refusal direction
+// is reachable in a test with no cosign installed and no AWS credentials — which
+// is the state of CI, and was the reason this branch had 0.0% coverage while
+// being the most security-relevant one in the agent (#56).
+type verifierPrereqs struct {
+	// lookPath resolves an executable, like exec.LookPath.
+	lookPath func(file string) (string, error)
+	// fetchKey returns a local path to the cosign public key, or "" on failure.
+	fetchKey func(ctx context.Context) string
+	// allowUnverified is the operator's explicit opt-out. False means refuse.
+	allowUnverified bool
 }
 
+// productionPrereqs wires the real prerequisites. getenv is a parameter so the
+// default this function produces — including the closed default for the opt-out
+// — is observable in a test rather than only by reading it.
+func productionPrereqs(getenv func(string) string) verifierPrereqs {
+	return verifierPrereqs{
+		lookPath:        exec.LookPath,
+		fetchKey:        fetchPublicKey,
+		allowUnverified: allowUnverified(getenv),
+	}
+}
+
+// allowUnverified reports whether the operator has explicitly opted out of
+// authenticity verification.
+//
+// Defaults to false for every input, including unset, empty, and unrecognised
+// values. A typo in the variable name or its value leaves verification on: the
+// direction of error for a security default is towards refusing.
+func allowUnverified(getenv func(string) string) bool {
+	switch strings.ToLower(strings.TrimSpace(getenv(allowUnverifiedEnv))) {
+	case "1", "true", "yes", "on":
+		return true
+	default:
+		return false
+	}
+}
+
+// resolveVerifier returns the verifier the agent should boot with.
+//
+// A non-nil error means the boot must stop. A nil verifier with a nil error
+// means the operator explicitly opted out — that pairing is the *only* way a
+// nil verifier is returned, and it is the invariant this function exists to
+// hold. Previously a nil verifier was returned on any prerequisite failure,
+// which made "cosign is not installed" and "the operator accepted the risk"
+// the same value.
+//
+// Why refusing is right here even though it costs boot availability: the agent
+// still checks SHA-256 against the lockfile either way, so what is lost when
+// verification is skipped is not integrity but *authenticity* — the lockfile
+// stops being evidence about who produced the layers, and a lockfile is exactly
+// the artifact an attacker would supply. Degrading that silently trades a
+// property nobody can observe was lost.
+func resolveVerifier(ctx context.Context, p verifierPrereqs) (trust.Verifier, error) {
+	v, err := newCosignVerifier(ctx, p)
+	if err == nil {
+		return v, nil
+	}
+	if !p.allowUnverified {
+		return nil, fmt.Errorf("%w (set %s=1 to boot without authenticity verification)", err, allowUnverifiedEnv)
+	}
+	log.Printf("strata-agent: WARNING: %s is set: booting WITHOUT authenticity verification (%v). "+
+		"SHA-256 integrity against the lockfile is still enforced; layer authorship is not checked.",
+		allowUnverifiedEnv, err)
+	return nil, nil
+}
+
+// newCosignVerifier builds a cosign verifier from the available prerequisites.
+//
+// It never returns a nil Verifier with a nil error: either both prerequisites
+// are present and a verifier is returned, or the missing one is named in an
+// error. Returning nil for "cannot verify" is the fail-open this replaces — and
+// it is not even a skip, since internal/trust/verify.go calls v.Verify
+// unconditionally.
+//
+// Prerequisites are reported in the order they are checked, and cosign comes
+// first deliberately: without the binary the key is useless, so naming the key
+// fetch when both are missing would send an operator to the wrong problem.
+func newCosignVerifier(ctx context.Context, p verifierPrereqs) (trust.Verifier, error) {
+	if _, err := p.lookPath("cosign"); err != nil {
+		return nil, fmt.Errorf("cosign not found on PATH: layer signatures cannot be verified: %w", err)
+	}
+	keyPath := p.fetchKey(ctx)
+	if keyPath == "" {
+		return nil, fmt.Errorf("could not fetch the cosign public key from s3://%s/%s: layer signatures cannot be verified",
+			registryBucket(), publicKeyObject)
+	}
+	return &trust.CosignVerifier{KeyRef: keyPath}, nil
+}
+
+// publicKeyObject is the registry key holding the Strata cosign public key.
+// Named here rather than inside fetchPublicKey so a refusal can tell the
+// operator which object could not be read.
+const publicKeyObject = "build/keys/cosign.pub"
+
 // fetchPublicKey downloads the Strata cosign public key from S3 to a temp
-// file and returns its path. Returns "" on any error (non-fatal).
+// file and returns its path. Returns "" on any error.
+//
+// The "" return is no longer a graceful degradation: its caller turns it into a
+// refusal to boot unless the operator has opted out. This function stays at 0.0%
+// coverage — it is an S3 GetObject against the registry and CI has no AWS
+// credentials by design — which is why the decision it feeds was moved out of it.
 func fetchPublicKey(ctx context.Context) string {
 	cfg, err := awsconfig.LoadDefaultConfig(ctx)
 	if err != nil {
@@ -100,10 +215,9 @@ func fetchPublicKey(ctx context.Context) string {
 	}
 	s3Client := s3.NewFromConfig(cfg)
 
-	const keyObject = "build/keys/cosign.pub"
 	out, err := s3Client.GetObject(ctx, &s3.GetObjectInput{
 		Bucket: aws.String(registryBucket()),
-		Key:    aws.String(keyObject),
+		Key:    aws.String(publicKeyObject),
 	})
 	if err != nil {
 		log.Printf("strata-agent: fetchPublicKey: %v", err)

